File: server/agent/schema.py
from typing import Any, Literal
from pydantic import BaseModel, Field, ConfigDict


class WidgetNode(BaseModel):
    """One node in the layout tree. Recursive."""
    model_config = ConfigDict(extra="allow")
    type: str
    id: str | None = None
    props: dict[str, Any] | None = None
    bindings: dict[str, str] | None = None
    children: list["WidgetNode"] | None = None


class ToolParameters(BaseModel):
    type: Literal["object"]
    properties: dict[str, Any]
    required: list[str] | None = None


class ToolImplementation(BaseModel):
    """A JSON-Patch template that the server applies when the tool is called.

    Each patch object can reference call arguments via ``"{{argname}}"``
    placeholder strings, which the server substitutes at call time. Example:

        [{"op": "add", "path": "/runs/-", "value": {
            "date": "{{date}}", "miles": "{{miles}}", "pace": "{{pace}}",
            "x": "{{date}}", "y": "{{miles}}"
        }}]
    """
    type: Literal["patch"] = "patch"
    patches: list[dict[str, Any]]


class ToolDeclaration(BaseModel):
    name: str
    description: str
    parameters: ToolParameters
    implementation: ToolImplementation


class BootstrapResponse(BaseModel):
    name: str = Field(min_length=1, max_length=40)
    icon: str = Field(min_length=1, max_length=8)
    appType: Literal["marathon", "trip", "jobs", "generic"]
    tree: WidgetNode
    data: dict[str, Any]
    tools: list[ToolDeclaration] = Field(min_length=1)
    backendCode: str | None = None


WidgetNode.model_rebuild()

File: server/agent/schema.ts
import { z } from "zod";

// One node in the layout tree. Recursive via z.lazy.
export const widgetSchema: z.ZodType<unknown> = z.lazy(() =>
  z.object({
    type: z.string(),
    id: z.string().optional(),
    props: z.record(z.unknown()).optional(),
    bindings: z.record(z.string()).optional(),
    children: z.array(widgetSchema).optional(),
  })
);

export const toolDeclarationSchema = z.object({
  name: z.string(),
  description: z.string(),
  parameters: z.object({
    type: z.literal("object"),
    properties: z.record(z.unknown()),
    required: z.array(z.string()).optional(),
  }),
  handler: z.string(),
});

export const bootstrapResponseSchema = z.object({
  name: z.string().min(1).max(40),
  icon: z.string().min(1).max(8),
  appType: z.enum(["marathon", "trip", "jobs", "generic"]),
  tree: widgetSchema,
  data: z.record(z.unknown()),
  tools: z.array(toolDeclarationSchema).min(1),
  backendCode: z.string().nullable(),
});

export type BootstrapResponse = z.infer<typeof bootstrapResponseSchema>;
